Fixed counting of unbound PVCs towards limit of attached volumes.

jsafrane opened the pull request:
Count unbound PVCs to the limit of attached volumes to a node. 

When MaxPDVolumeCountPredicate is in doubt (e.g. PVC or PV is missing), it assumes the volume is attached. It should assume the same when it encounters an unbound PVC. In any case, it should not return an error, it would stop scheduling all pods with a PVC.

Fixes: #53134

```release-note
NONE
```

resouer commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -228,17 +228,21 @@ func (c *MaxPDVolumeCountChecker) filterVolumes(volumes []v1.Volume, namespace s
+				// the PV towards the PV limit
+				glog.V(4).Infof("PVC %s/%s is not bound, assuming PVC matches predicate when counting limits", namespace, pvcName)
+				source := rand.NewSource(time.Now().UnixNano())
+				generatedID := "unboundPVC" + strconv.Itoa(rand.New(source).Intn(1000000))
```
May be it's better to use exiting util for rand name here?

jsafrane replied:
This code is copied in all error paths in `filterVolumes`, I don't wan't a different behavior in one of them.

dixudx commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -218,7 +218,7 @@ func (c *MaxPDVolumeCountChecker) filterVolumes(volumes []v1.Volume, namespace s
 			}
 			pvc, err := c.pvcInfo.GetPersistentVolumeClaimInfo(namespace, pvcName)
-			if err != nil {
+			if err != nil || pvc == nil {
```
Should not `pvc` be `nil` when `err != nil`?

jsafrane replied:
There was `pvc == nil` check few lines below, I just moved it here to keep the same behavior as the original predicate.

bsalamat commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -218,7 +218,7 @@ func (c *MaxPDVolumeCountChecker) filterVolumes(volumes []v1.Volume, namespace s
 			}
 			pvc, err := c.pvcInfo.GetPersistentVolumeClaimInfo(namespace, pvcName)
-			if err != nil {
+			if err != nil || pvc == nil {
```
I don't understand this. Why should we return nil (no error) when the PVC is not found?

cc/ @msau42

msau42 replied:
This predicate is counting the number of PVs that are running on a single Node across all namespaces and Pods.

Users can delete PVC objects even when a Pod is still running and using it.  So even though the PVC object doesn't exist anymore, the volume could still be mounted and used on the node, so we should count it since the Pod is referencing it.

msau42 commented on plugin/pkg/scheduler/algorithm/predicates/predicates_test.go:
```diff
@@ -1823,6 +1836,20 @@ func TestEBSVolumeCountConflicts(t *testing.T) {
+			existingPods: []*v1.Pod{oneVolPod, unboundPVCPod},
+			maxVols:      3,
+			fits:         true,
+			test:         "pod with unbound PVC is counted towards the PV limit",
```
what about a test case if pvc doesn't exist?

jsafrane replied:
there already is a test for that, look for "pod with missing PVC is counted towards the PV limit" above.

msau42 replied:
How about multiple PV objects not existing in the same pod and different pods?

jsafrane replied:
Added a test with two pods using the same deleted PV (=counted once) and two pods using a different deleted PV (=counted twice).

jingxu97 commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -249,10 +253,6 @@ func (c *MaxPDVolumeCountChecker) filterVolumes(volumes []v1.Volume, namespace s
 				return nil
```
why it returns immediately whenever there is error instead of continue to check the following volumes

jsafrane replied:
That's actually a good question. `return` was there even before this PR and it looks wrong. I'll rework the PR to make it robust when a pod uses two deleted PVCs (counted as two) and two pods using the same deleted PVC (counted as one).

jsafrane committed:
Fixed counting of unbound PVCs towards limit of attached volumes.

There are two ways how a scheduled pod can get its PVCs unbound:
- admin forcefuly unbinds it
- user deletes original PVC that was bound when the pod was scheduled and
  creates a new one with the same name that does not get bound from some
  reason.

In both cases we don't know where the original PVC pointed at and if we
should account it to the limit of attached AWS EBS / GCE PDs etc.

The common pattern here is to count it in when in doubt.

jsafrane commented:
Reworked whole `filterVolumes` to count all volumes properly and don't return on first unbound / deleted PVC.

One deleted or unbound PVC is now counted once when it's used in multiple pods. Two different deleted or unbound PVCs are counted as two volumes, even when used in a single pod.

jsafrane commented:
/retest

msau42 commented:
/lgtm

bsalamat commented:
/approve

Thanks, @jsafrane and @msau42!

liggitt commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -217,40 +215,37 @@ func (c *MaxPDVolumeCountChecker) filterVolumes(volumes []v1.Volume, namespace s
+
+			// Until we know real ID of the volume use namespace/pvcName as substitute
+			// With a random prefix so it can't conflict with existing volume ID.
+			pvId := fmt.Sprintf("%s-%s/%s", randomPrefix, namespace, pvcName)
```
Why not use an incrementing 64 bit counter here

liggitt commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -269,8 +264,15 @@ func (c *MaxPDVolumeCountChecker) predicate(pod *v1.Pod, meta algorithm.Predicat
+	// is random to avoid conflicts with real volume IDs and it needs to be
+	// stable in whole predicate() call so a deleted PVC used by two pods is
+	// counted as one volume and not as two.
+	randomPrefix := rand.String(32)
```
also, this makes an expensive call even when it is not needed, right?
